src/main/java/net/neoforged/neoforge/fluids/crafting/UnitAmountFluidIngredient.java

```diff
@@ -0,0 +1,288 @@
+        return new UnitAmountFluidIngredient(fluid, List.of(Map.of(unit, amount)));
+    }
+
+    /**
+     * Helper method to create a simple unit-based ingredient that matches a single fluid.
+     */
+    public static UnitAmountFluidIngredient of(Fluid fluid, int amount, FluidUnit unit) {
+        return new UnitAmountFluidIngredient(FluidIngredient.of(fluid), List.of(Map.of(unit, amount)));
+    }
+
+    /**
+     * Helper method to create a simple unit-based ingredient that matches a single fluid.
+     */
+    public static UnitAmountFluidIngredient of(TagKey<Fluid> tag, int amount, FluidUnit unit) {
+        return new UnitAmountFluidIngredient(FluidIngredient.tag(tag), List.of(Map.of(unit, amount)));
+    }
+
+    /**
+     * Factory method for creating any ingredient.
+     */
+    @SafeVarargs
+    public static UnitAmountFluidIngredient of(FluidIngredient ingredient, Map<FluidUnit, Integer>... amounts) {
+        if (amounts.length == 0)
+            return new UnitAmountFluidIngredient(ingredient, DEFAULT_AMOUNT);
+        else
+            return new UnitAmountFluidIngredient(ingredient, Arrays.stream(amounts)
+                    .map(UnitAmountFluidIngredient::validateNotEmpty)
+                    .map(DataResult::getOrThrow).toList());
+    }
+
+    private final FluidIngredient ingredient;
+    private final List<Map<FluidUnit, Integer>> amounts;
+
+    @Nullable
+    private FluidStack[] cachedStacks;
+
+    private UnitAmountFluidIngredient(FluidIngredient ingredient, List<Map<FluidUnit, Integer>> amounts) {
+        this.ingredient = ingredient;
+        this.amounts = amounts;
+    }
+
+    /**
+     * The unsized fluid ingredient this ingredient is based on
+     */
+    public FluidIngredient ingredient() {
+        return ingredient;
+    }
+
+    /**
+     * The list of unit-based amounts associated with this recipe.
+     * Each element of the list is checked in turn, the first one that has all its units defined
+     * in the fluid's unit spec is used. If none of them qualify, the ingredient does not match.
+     */
+    public List<Map<FluidUnit, Integer>> amounts() {
+        return amounts;
+    }
+
+    /**
+     * Get the quantity in millibuckets associated with the recipe's amount
+     * 
+     * @return The millibucket quantity for the first amount that has all its units defined in the fluid spec
+     */
+    public Optional<Integer> amount(FluidStack stack) {
+        return amount(stack.getFluidType().getUnits(), amounts);
+    }
+
+    /**
+     * Get the quantity in millibuckets associated with a given unit-based amount
+     * 
+     * @param fluidUnits The unit spec of the fluid
+     * @param amounts    List of amounts as specified in the recipe's amount fluid
+     * @return The millibucket quantity for the first amount that has all its units defined in the fluid spec
+     */
+    public static Optional<Integer> amount(FluidUnitSpec fluidUnits, List<Map<FluidUnit, Integer>> amounts) {
```

Comment:
Do I understand this method correctly: 

If i ask for a FluidUnit of Bird, that by definition contains 123 mB of fluid, yet the Fluid does not support that FluidUnit then an empty Optional, or at least an Optional with 0 mB is returned?

Because we loop through the requested amounts: 1 Bird.
The FluidTypes FluidUnitSpec is querried for the value. Which checks wether it is a bucket, or an Mb (neither of which it is), then goes to its known units map, which has no Bird. And returns an empty optional. This breaks the inner loop, and because I am only requesting 1 Bird that completes the loop over amounts. Returning an empty optional.

Correct?

Reply:
Yes that's intentional behaviour. Fluid units are only valid for fluids that explicitly support them, and in particular it's valid for the same unit to have a different size for different fluid types. If a fluid doesn't declare support, it's saying that it cannot be broken down into "Bird" or whatever. This allows one fluid to break down to have 9 ingots to a block, while a different one has 4 ingots to a block.

Reply:
Except that this is terrible for compatibility with mods that add custom containers, like custom bottles or other systems.

Reply:
It's a data map, so the mod could just define the sizes for the fluids it wants to support, and players can add others as necessary. Ideally containers should use a multiple of one of the common units, and mods adding fluids should define sizes for those whenever they make sense. The point is to make it possible to have a future where the millibucket size is an internal thing to any given fluid, so different ones can use different standards depending on how they need to be broken down. 

Alternatively if you want your container to support anything, you could inspect the units defined for a fluid, then adjust the max capacity to fit whatever limits you want. Or stick to buckets, since "1 block full of fluid" is universally applicable.

Reply:
Neither of these options are something I am very comfortable with adding as an API.
The current design of the API allows any fluid to be stored in any container with any shape or size. Weither that is mB, Ingots or Birds. It does not matter.

It was specifically picked to allow for the greatest compatibility between mods, which was always the aim of this project.

The idea that a fluid can define its own mB is simply not really a good idea. The mB is defined not based on the fluid, but simply by the fact that it is 1/1000 of the volume of a single block of the liquid.

What you want is the ability of the fluid to say: Well if you pick me up, then I am 800 mB instead of a 1000, which is a different API to discuss.

As it stands right now that method that I referenced, combined with the the Ingredient is especially problematic. Because there is nothing in this API that prevents me from putting fluid A into container B that is of unit Bird. Presenting 1 Bird of Fluid A to the recipe, which in mB would be acceptable since my recipe would take for example 50 mB leaving, the remainder container behind with 73 mB (aka 73/123 Birds) (that is fine, our current APIs already support this scenario). Today this is completely functional and working, yet with this API that recipe would not match, since the fluid does not know my generic Bird unit (unless I datapack it into everything, which is simply a bad API design to start with, yes it works for individual units that are added to individual fluids, but not for generic units, like buckets, bottles, and in this case Birds).

All in all, I think this API focuses too much on the abilities of the fluid, not enough on the abilities of the container it is put it. And due to the dynamicness of containers, I think that makes them require a more complex API. A good start would be to not have the datamap in the first place.

But have a conversion system, that allows the conversion of one unit into another unit of fluid. Regardless of fluid. That is how it works in the real world too. We don't redefine the KG or the Meter because we are measuring them against a different medium (yes i know there are miles, but lets not go there).

Reply:
> Except that this is terrible for compatibility with mods that add custom containers, like custom bottles or other systems.

No its not. This is fantastic for compatibility for that. You either

* Use millibuckets. No one is stopping you.
* Define a custom unit with some fallbacks to common units; this is impossible outside of this API.

Like seriously, why do so many people assume that you can no longer use millibuckets? Its literally defined in the list of common units! This API does not limit you in any way whatsoever, it just adds new options.


> The current design of the API allows any fluid to be stored in any container with any shape or size. Weither that is mB, Ingots or Birds. It does not matter.
> it was specifically picked to allow for the greatest compatibility between mods, which was always the aim of this project.

Wait, are you talking about the Forge 1000mb API? Is this a joke?

The current NeoForge API was chosen on a whim probably 10 years ago because that is what the author of the API used in his mods. At the time we had no idea what people would need.

We have come to discover 1000mb is absolutely horrible as it is incompstible with vanilla.

> The idea that a fluid can define its own mB is simply not really a good idea. The mB is defined not based on the fluid, but simply by the fact that it is 1/1000 of the volume of a single block of the liquid.

Millibuckets are the one unit that is fixed at this stage. No one is redefining millibuckets.

Rather, the concept of a unit of fluid is just being softcoded. Right now we assume that all fluids divide into millibuckets and that is fixed. In the future we might decide all fluids divide into droplets, or some other minimum unit. Alternatively, we might decide each fluid decides its own divisions. Neither of those are this API. But they do emphasize one important fact you are missing: millibuckets are a stupid unit that no mod really needs. We all use some division of millibuckets which we give a name.

Supposing one of those future paths is taken and millibucket is no longer the min unit, nothing stops you from doing something dumb and making it not 1/1000 of a bucket. But nothing stops you from adding a steak to the iron ingot tag either, we should code our APIs to be useful, not artificially restrict them to prevent people from being dumb.

> What you want is the ability of the fluid to say: Well if you pick me up, then I am 800 mB instead of a 1000, which is a different API to discuss.

Who said anything about pickup? This is not a change to buckets.

The idea here is just recipe input matching. My recipe wants 1 ingot, I don't care if 1 ingot is 100mb or 90mb, your fluid decides that.

If we want to use the API on buckets or cauldrons or any of that nonsense in the future we can, but deciding against using it there does not make it useless; this is being pitched for its benefits outside those areas.

> As it stands right now that method that I referenced, combined with the the Ingredient is especially problematic. Because there is nothing in this API that prevents me from putting fluid A into container B that is of unit Bird. Presenting 1 Bird of Fluid A to the recipe, which in mB would be acceptable since my recipe would take for example 50 mB leaving, the remainder container behind with 73 mB (aka 73/123 Birds) (that is fine, our current APIs already support this scenario). Today this is completely functional and working, yet with this API that recipe would not match, since the fluid does not know my generic Bird unit (unless I datapack it into everything, which is simply a bad API design to start with, yes it works for individual units that are added to individual fluids, but not for generic units, like buckets, bottles, and in this case Birds).

***FALLBACKS***, that is why its a list of maps and not just a map.

Again, you have several ***incredibly simple*** solutions to this problem you are overcomplicating.

* Option 1: I want my container to hold *anything*. I define its capacity in terms of millibuckets. I don't need to use this API.
* Option 2: I want my container to hold 1 ingot, I define it as taking 1 ingot using this API. Does not matter how much 1 ingot is, and I don't support it if the fluid lacks ingots.
* Option 3: fallbacks! I make my container hold 1 ingot, or 1 gem, or 90mb. Now it works for any fluid again, but lets fluids following different standards use their own ingots.

You don't have to datapack every fluid if you just use the fallbacks clearly shown in this function you are commenting on.

> All in all, I think this API focuses too much on the abilities of the fluid, not enough on the abilities of the container it is put it. And due to the dynamicness of containers, I think that makes them require a more complex API. A good start would be to not have the datamap in the first place.

What is missing in containers? I make my container hold 1000mb. Or 10 ingots. Or whatever other size I want. Containers are dead simple.

I could make my container hold in increments of ingots, or I could make it increments of millibuckets like we already have and just set its capacity to be a unit lookup. That would just be overriding a single method that already exists in a fluid handler, the fluid stack sensitive capacity.

> But have a conversion system, that allows the conversion of one unit into another unit of fluid. Regardless of fluid. That is how it works in the real world too. We don't redefine the KG or the Meter because we are measuring them against a different medium (yes i know there are miles, but lets not go there).

In the real world, different fluids have different densities, so while a volume measurement might be the same, units such as moles vary based on what you are measuring. Or weight units, or mass. But real world does not matter here, because Minecraft is not the real world.

Our goal is constructing a useful API, not a physics simulator.

Reply:
The issue is that you started out with 
> If i ask for a FluidUnit of Bird, that by definition contains 123 mB of fluid

Which is missing the whole point of the API. If you make a fluid unit, by definition it _doesn't_ always contain 123 mB of fluid. Cause fluids tell you how much a unit is. If you need 123 mB of fluid -- you shouldn't be using the fluid unit system!

Reply:
I seem to have had this PR upside down.

The Idea is for consumers of fluids not containers of fluids.
That makes more sense.

So that the consumer is able to figure out how much fluid he or she needs to extract from a container in the base unit (mB) given a whole larger unit that he needs, like ingots, nuggets, Birds, etc.

So different conversation. Sorry, the PR description, and the Gist does not seem to make that very clear. Especially since others keep talking about mB and potentially redefining that. 

Okey problem solved, thank you for the clear up.

Reply:
Part of that might just be this is a pretty base API which could be expanded in those directions based on what the community wants.

The root problem is if we just change the bucket divisions from 1000 to some other number, that is going to be confusing for both players and modders as all the numbers change. The minimal fix is a way to render units in the tooltip so at least players are not confused (they had ingots before and they have ingots now, does not matter under the hood its a different number).

That concept though has a lot of potential applications; right now we are focused on consumers as thats the minimal change needed to change the bucket divisions or just make 1000mb nicer to work with. In the future based on what the community wants we could consider concepts such as fluid transfer or fluid storage; its all comes down to which path people wish to take: leave buckets at 1000mb, change it to another unit like droplets, or let it be defined per fluid.

Reply:
Yeah, this PR is just poorly documented, and the gist really does not help when trying to understand the actual API.

Reply:
Correct - in the wider context of the ingredient type, the recipe will not match if it requests units that the actual fluid present in the recipe input doesn't support, even if it otherwise matches (in case, say, i have a recipe to make an iron ingot from one ingot amount of molten iron, but someone has added a fluid to the molten iron tag that does not have an ingot unit)

> FluidUnit of Bird, that by definition contains 123 mB of fluid

This statement seems confused. Each fluid has its own definitions for units. There is no "by definition" without the context of a specific fluid. One mod's molten iron fluid may have 90 mb per ingot, another may have 100 or 144.